javalib/src/main/scala/java/net/InetAddress.scala

```diff
@@ -212,32 +215,63 @@ object InetAddress {
      * to leave the host field blank/empty.
      */
     val effectiveHost = if (isNumeric) null else host
-    val ipAddress = SocketHelpers.sockaddrToByteArray(addrinfoP.ai_addr)
-    if (ipAddress.length == 4) {
-      new Inet4Address(ipAddress, effectiveHost)
-    } else {
+
+    if (addrinfoP.ai_family == AF_INET) {
+      new Inet4Address(addrinfoToByteArray(addrinfoP), effectiveHost)
+    } else if (addrinfoP.ai_family == AF_INET6) {
       val addr = addrinfoP.ai_addr.asInstanceOf[Ptr[sockaddr_in6]]
-      /* Yes, Java specifies Int for scope_id in a way which disallows
-       * some values POSIX/IEEE/IETF allows.
-       */
-      val scope_id = addr.sin6_scope_id.toInt
+      val addrBytes = addr.sin6_addr.at1.at(0).asInstanceOf[Ptr[Byte]]
 
-      val zoneIdent = {
-        val ifIndex = host.indexOf('%')
-        val ifNameStart = ifIndex + 1
-        if ((ifIndex < 0) || (ifNameStart >= host.length)) ""
-        else host.substring(ifNameStart)
-      }
+      // Scala JVM down-converts even when preferIPv6Addresses is "true"
+      if (isIPv4MappedAddress(addrBytes)) {
+        new Inet4Address(extractIP4Bytes(addrBytes), effectiveHost)
```

Comment:
`addrinfoToByteArray` already handles this. We can call the function and check the length of the returned array instead. This would avoid to test for IPv4 mapped address twice. WDYT ?

Reply:
re: // Scala JVM down-converts even when preferIPv6Addresses is "true"

I am not sure if I still believe that comment.  I will have to check it out.

I believe that JVM will __display_ an IPv4 mapped IPv6 address.  My reading 
since that comment was written makes me think that the JVM carries IPv4
mapped IPv6 addresses internally as IPv6 and not IPv4. 

I just do not remember if SN allows parsing IPv4 mapped IPv6 addresses, potentially with
an interface_index/scope_id on IPv4 only systems. My guess would be SN yes, JVM no.

On systems where the network stack is dual protocol IPv4/IPv6, the current internal is probably
invisible if the other end of the connection is also dual stack (and has chosen to listen on both
protocols, the default). When SN downconverts internally (i.e. early, just after parse) and the
remote end is listening on only IPv6, there might be a defect.

I am told, recently, that not all systems are "dual stack". Specifically I am told that FreeBSD and NetBSD
like to maintain independent IPv4 & IPv6 stacks (and can configure either or both off).  I believe,
but have not tested, that the current 'downconvert early' practice will break there. 

I wasn't doing anything in Q1, Q2, and possibly Q3 2024 anyway.